REGRESSION (265825@main): [iOS] europa.eu: Scroll position jitters when scrolling over interactive map

whsieh opened the pull request:
#### 8612f9f7a3cda362a6280a9ef0e0e687f4d360c1
<pre>
REGRESSION (265825@main): [iOS] europa.eu: Scroll position jitters when scrolling over interactive map
<a href="https://bugs.webkit.org/show_bug.cgi?id=282933">https://bugs.webkit.org/show_bug.cgi?id=282933</a>
<a href="https://rdar.apple.com/112427506">rdar://112427506</a>

Reviewed by Abrar Rahman Protyasha.

The changes in 265825@main exacerbated an existing bug, regarding how our touch event handling logic
propagates the location of the user&apos;s touches from the UI process to the web process; this results
in scrolling on the immigration portal on europa.eu becoming extremely jittery (bouncing between the
minimum and maximum scroll positions in the main document) as the user performs an upwards swipe
over the interactive map to scroll down.

The interactive map on europa.eu intercepts and prevents all `touchstart` and `touchmove` events
that are dispatched over it. When the user attempts to scroll the page by swiping over the element
and `touchmove` is dispatched, their script attempts to adjust `pageYOffset` such that the location
of the user&apos;s touch relative to the document remains the same, before and after handling the
`touchmove` delta. In other words, the `touchmove` event listener simulates scrolling by keeping a
point in content coordinates fixed underneath the user&apos;s finger as they pan.

However, this doesn&apos;t quite work well in practice, especially after the changes in 265825@main
elided an extra CA commit that would otherwise more aggressively sync programmatic scrolling to UI-
side scroll content offset. That&apos;s because touch locations are computed in `WKContentView`
coordinates in the UI process (when receiving gesture recognizer subclass method calls from UIKit),
and then dispatched to the web process where they&apos;re asynchronously queued and delivered to event
listeners for processing. Throughout this process, the root view location of the touch remains
unchanged since it was computed in the UI process, which means that it ignores any changes to the
scroll position of the page that happened after the gesture was recognized. This is particularly
disastrous when combined with europa.eu&apos;s touch event handling logic, which constantly shifts the
scroll position when receiving `touchmove` events — the result is that the page can be scrolled
programmatically, but the root view coordinates of subsequent `touchmove` touch locations don&apos;t
reflect this scrolling at all, causing the event listener to try and move the `pageYOffset` in the
*opposite* direction to compensate. This error then compounds very quickly, causing the scroll
position to jump between min and max scroll positions on the page.

To fix this, we make some adjustments to propagate the user&apos;s touch location relative to the
viewport (unobscured content rect), and only map from the viewport back to root view coordinates at
the last possible moment before dispatching the event by shifting the touch delta back up, using the
main frame&apos;s current `contentsScrollPosition()`.

See below for more details.

* LayoutTests/fast/events/touch/ios/programmatic-scrolling-on-touchmove-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/programmatic-scrolling-on-touchmove.html: Added.

Add a layout test to exercise the change, by swiping up in a reduced test case that emulates
europa.eu&apos;s logic for scrolling in response to `touchmove`, and verifying that the scroll position
always stays within reasonable limits. Without the fixes in this bug, the `pageYOffset` ends up
bouncing between negative values and values of up to 2000-3000.

* Source/WebCore/dom/ios/MouseEventIOS.cpp:
(WebCore::MouseEvent::create):
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::dispatchEventForTouchAtIndex):
* Source/WebCore/platform/ios/PlatformEventFactoryIOS.mm:
(WebCore::PlatformTouchPointBuilder::PlatformTouchPointBuilder):
(WebCore::PlatformTouchEventBuilder::PlatformTouchEventBuilder):

Pass `locationInViewport` through to `PlatformTouchPoint` on construction. This is optional since
we currently don&apos;t compute this for `UIWebView` (WebKitLegacy).

* Source/WebCore/platform/ios/ScrollAnimatorIOS.mm:

Adjust call sites, since `touchLocationAtIndex` is now renamed to `touchLocationInRootViewAtIndex`.

(WebCore::ScrollAnimatorIOS::handleTouchEvent):
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformTouchPoint::WebKit2PlatformTouchPoint):
* Source/WebKit/Shared/WebTouchEvent.h:
(WebKit::WebPlatformTouchPoint::WebPlatformTouchPoint):
(WebKit::WebPlatformTouchPoint::locationInRootView const):
(WebKit::WebPlatformTouchPoint::locationInViewport const):
(WebKit::WebPlatformTouchPoint::location const): Deleted.

Rename `location` to `locationInRootView` (along with other related members), and plumb a new
`locationInViewport` point alongside the touch point.

* Source/WebKit/Shared/ios/NativeWebTouchEventIOS.mm:
(WebKit::NativeWebTouchEvent::extractWebTouchPoints):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateTouchEventTracking):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:

Refactor `WKTouchEventsGestureRecognizer` to just directly take a `WKContentView`. Now that the
touch events gesture no longer needs to be generic for both WebKitLegacy and modern WebKit, there&apos;s
no benefit to maintaining a separate, WebKit-internal `WKTouchEventsGestureRecognizerDelegate` that
is only implemented by the content view.

This allows us to remove some unused code and unnecessary logic (such as the touch target and action
selector, as well as `-isAnyTouchOverActiveArea:`), but importantly allows us to call into the web
view to compute the unobscured content offset when determining the viewport-relative location for
each `WKTouchPoint`.

(-[WKContentView setUpInteraction]):
(-[WKContentView _touchEventsRecognized]):
(-[WKContentView _handleTouchActionsForTouchEvent:]):
(-[WKContentView _shouldIgnoreTouchEvent:]):
(-[WKContentView _touchEventsRecognized:]): Deleted.
(-[WKContentView gestureRecognizer:shouldIgnoreTouchEvent:]): Deleted.
(-[WKContentView isAnyTouchOverActiveArea:]): Deleted.
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.h:
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm:
(-[WKTouchEventsGestureRecognizer initWithContentView:]):
(mapRootViewToViewport):

Add a helper function to unadjust the given point in root view coordinates by the current unobscured
content offset.

(-[WKTouchEventsGestureRecognizer _touchEventForTouch:]):
(-[WKTouchEventsGestureRecognizer _recordTouches:type:coalescedTouches:predictedTouches:]):

Populate the new `locationInViewport` member on each touch point.

(-[WKTouchEventsGestureRecognizer performAction]):
(-[WKTouchEventsGestureRecognizer touchesBegan:withEvent:]):
(-[WKTouchEventsGestureRecognizer initWithTarget:action:touchDelegate:]): Deleted.
* Tools/TestWebKitAPI/Tests/ios/TouchEventTests.mm:

Canonical link: <a href="https://commits.webkit.org/286448@main">https://commits.webkit.org/286448@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6471a2cbe019fd75f43a39906ac57c1fdd28dc43

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76011 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55040 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28910 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80508 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27277 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78127 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64182 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3336 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59598 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17759 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79078 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49487 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65282 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39956 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46886 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22766 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25604 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68002 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23103 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81972 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3380 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2162 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67829 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3534 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65250 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67138 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16727 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11091 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9213 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3330 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6136 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3351 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4289 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3358 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on LayoutTests/fast/events/touch/ios/programmatic-scrolling-on-touchmove.html:
```diff
@@ -0,0 +1,63 @@
+
+    container.addEventListener("touchmove", (event) => {
+        const position = Math.round(event.touches[0].pageY);
+        const delta = Math.max(position, startPosition) - Math.min(position, startPosition)
```
nit: `Math.abs(position - startPosition)`?

whsieh replied:
Ah, I forgot to remove this `delta` altogether since it's now unused. Will do that…

whsieh committed:
REGRESSION (265825@main): [iOS] europa.eu: Scroll position jitters when scrolling over interactive map
https://bugs.webkit.org/show_bug.cgi?id=282933
rdar://112427506

Reviewed by Abrar Rahman Protyasha.

The changes in 265825@main exacerbated an existing bug, regarding how our touch event handling logic
propagates the location of the user's touches from the UI process to the web process; this results
in scrolling on the immigration portal on europa.eu becoming extremely jittery (bouncing between the
minimum and maximum scroll positions in the main document) as the user performs an upwards swipe
over the interactive map to scroll down.

The interactive map on europa.eu intercepts and prevents all `touchstart` and `touchmove` events
that are dispatched over it. When the user attempts to scroll the page by swiping over the element
and `touchmove` is dispatched, their script attempts to adjust `pageYOffset` such that the location
of the user's touch relative to the document remains the same, before and after handling the
`touchmove` delta. In other words, the `touchmove` event listener simulates scrolling by keeping a
point in content coordinates fixed underneath the user's finger as they pan.

However, this doesn't quite work well in practice, especially after the changes in 265825@main
elided an extra CA commit that would otherwise more aggressively sync programmatic scrolling to UI-
side scroll content offset. That's because touch locations are computed in `WKContentView`
coordinates in the UI process (when receiving gesture recognizer subclass method calls from UIKit),
and then dispatched to the web process where they're asynchronously queued and delivered to event
listeners for processing. Throughout this process, the root view location of the touch remains
unchanged since it was computed in the UI process, which means that it ignores any changes to the
scroll position of the page that happened after the gesture was recognized. This is particularly
disastrous when combined with europa.eu's touch event handling logic, which constantly shifts the
scroll position when receiving `touchmove` events — the result is that the page can be scrolled
programmatically, but the root view coordinates of subsequent `touchmove` touch locations don't
reflect this scrolling at all, causing the event listener to try and move the `pageYOffset` in the
*opposite* direction to compensate. This error then compounds very quickly, causing the scroll
position to jump between min and max scroll positions on the page.

To fix this, we make some adjustments to propagate the user's touch location relative to the
viewport (unobscured content rect), and only map from the viewport back to root view coordinates at
the last possible moment before dispatching the event by shifting the touch delta back up, using the
main frame's current `contentsScrollPosition()`.

See below for more details.

* LayoutTests/fast/events/touch/ios/programmatic-scrolling-on-touchmove-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/programmatic-scrolling-on-touchmove.html: Added.

Add a layout test to exercise the change, by swiping up in a reduced test case that emulates
europa.eu's logic for scrolling in response to `touchmove`, and verifying that the scroll position
always stays within reasonable limits. Without the fixes in this bug, the `pageYOffset` ends up
bouncing between negative values and values of up to 2000-3000.

* Source/WebCore/dom/ios/MouseEventIOS.cpp:
(WebCore::MouseEvent::create):
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::dispatchEventForTouchAtIndex):
* Source/WebCore/platform/ios/PlatformEventFactoryIOS.mm:
(WebCore::PlatformTouchPointBuilder::PlatformTouchPointBuilder):
(WebCore::PlatformTouchEventBuilder::PlatformTouchEventBuilder):

Pass `locationInViewport` through to `PlatformTouchPoint` on construction. This is optional since
we currently don't compute this for `UIWebView` (WebKitLegacy).

* Source/WebCore/platform/ios/ScrollAnimatorIOS.mm:

Adjust call sites, since `touchLocationAtIndex` is now renamed to `touchLocationInRootViewAtIndex`.

(WebCore::ScrollAnimatorIOS::handleTouchEvent):
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformTouchPoint::WebKit2PlatformTouchPoint):
* Source/WebKit/Shared/WebTouchEvent.h:
(WebKit::WebPlatformTouchPoint::WebPlatformTouchPoint):
(WebKit::WebPlatformTouchPoint::locationInRootView const):
(WebKit::WebPlatformTouchPoint::locationInViewport const):
(WebKit::WebPlatformTouchPoint::location const): Deleted.

Rename `location` to `locationInRootView` (along with other related members), and plumb a new
`locationInViewport` point alongside the touch point.

* Source/WebKit/Shared/ios/NativeWebTouchEventIOS.mm:
(WebKit::NativeWebTouchEvent::extractWebTouchPoints):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateTouchEventTracking):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:

Refactor `WKTouchEventsGestureRecognizer` to just directly take a `WKContentView`. Now that the
touch events gesture no longer needs to be generic for both WebKitLegacy and modern WebKit, there's
no benefit to maintaining a separate, WebKit-internal `WKTouchEventsGestureRecognizerDelegate` that
is only implemented by the content view.

This allows us to remove some unused code and unnecessary logic (such as the touch target and action
selector, as well as `-isAnyTouchOverActiveArea:`), but importantly allows us to call into the web
view to compute the unobscured content offset when determining the viewport-relative location for
each `WKTouchPoint`.

(-[WKContentView setUpInteraction]):
(-[WKContentView _touchEventsRecognized]):
(-[WKContentView _handleTouchActionsForTouchEvent:]):
(-[WKContentView _shouldIgnoreTouchEvent:]):
(-[WKContentView _touchEventsRecognized:]): Deleted.
(-[WKContentView gestureRecognizer:shouldIgnoreTouchEvent:]): Deleted.
(-[WKContentView isAnyTouchOverActiveArea:]): Deleted.
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.h:
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm:
(-[WKTouchEventsGestureRecognizer initWithContentView:]):
(mapRootViewToViewport):

Add a helper function to unadjust the given point in root view coordinates by the current unobscured
content offset.

(-[WKTouchEventsGestureRecognizer _touchEventForTouch:]):
(-[WKTouchEventsGestureRecognizer _recordTouches:type:coalescedTouches:predictedTouches:]):

Populate the new `locationInViewport` member on each touch point.

(-[WKTouchEventsGestureRecognizer performAction]):
(-[WKTouchEventsGestureRecognizer touchesBegan:withEvent:]):
(-[WKTouchEventsGestureRecognizer initWithTarget:action:touchDelegate:]): Deleted.
* Tools/TestWebKitAPI/Tests/ios/TouchEventTests.mm:

Canonical link: https://commits.webkit.org/286448@main
